File: about.iOS/about.GPUImage/CustomOutput/CustomOutput/ViewController.m
//
//  ViewController.m
//  CustomOutput
//
//  Created by burt on 3/30/16.
//  Copyright © 2016 burt. All rights reserved.
//

#import "ViewController.h"
#import <GPUImage/GPUImage.h>
#import "GPUImageSpriteOverlayOutput.h"
#import "GPUImageTempFilter.h"
#import <YYImage/YYImage.h>
@import GLKit;

@interface ViewController ()
@property(nonatomic, strong) GPUImageVideoCamera* videoCamera;
@property(nonatomic, strong) GPUImageView* videoOut;
@property(nonatomic, strong) GPUImageAlphaBlendFilter* middle;
@property(nonatomic, strong) GPUImageAlphaBlendFilter* final;
@property(nonatomic, strong) GPUImageMovie* movie;
@property(nonatomic, strong) GPUImageSpriteOverlayOutput* sprite;
@property(nonatomic, strong) GPUImageTempFilter *filter;

@property(nonatomic, strong) UIImage *image;
@property(nonatomic, strong) GLKTextureInfo *textureInfo;

@end

@implementation ViewController

- (void)viewDidLoad {

    [super viewDidLoad];


    static const GLfloat squareVertices[] = {
        -0.5f, -0.5f,
        0.5f, -0.5f,
        -0.5f,  0.5f,
        0.5f,  0.5f,
    };
    
    static const GLfloat spriteTextureCoordinates[] = {
        0.0f, 0.0f,
        1.0f, 0.0f,
        0.0f, 1.0f,
        1.0f, 1.0f,
    };
    
    [[GPUImageContext sharedImageProcessingContext] useAsCurrentContext];
    self.image = [UIImage imageNamed:@"heart.png"];
    CGImageRef cgImage = self.image.CGImage;
    NSError *err;
    
    self.textureInfo = [GLKTextureLoader textureWithCGImage:cgImage
                                                    options:nil
                                                      error:&err];
    
    self.videoOut = [[GPUImageView alloc] initWithFrame:self.view.bounds];
    [self.view addSubview:self.videoOut];
    
    self.videoCamera = [[GPUImageVideoCamera alloc] initWithSessionPreset:AVCaptureSessionPreset640x480 cameraPosition:AVCaptureDevicePositionFront];
    self.videoCamera.outputImageOrientation = UIInterfaceOrientationPortrait;
    self.videoCamera.horizontallyMirrorRearFacingCamera = YES;
    self.filter = [[GPUImageTempFilter alloc] init];
    
    __weak ViewController *weakSelf = self;
    self.filter.contextUpdateBlock = ^(GLint positionAttribute, GLint textureCoordinateAttribute, GLint inputTextureUniform) {
        glBindTexture(weakSelf.textureInfo.target, weakSelf.textureInfo.name);
        glUniform1i(inputTextureUniform, 2);
        glVertexAttribPointer(positionAttribute, 2, GL_FLOAT, 0, 0, squareVertices);
        glVertexAttribPointer(textureCoordinateAttribute, 2, GL_FLOAT, 0, 0, spriteTextureCoordinates);
    };
    
    [self.videoCamera addTarget:self.filter];
    [self.filter addTarget:self.videoOut];
    
    [self.videoCamera startCameraCapture];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

@end

File: about.GPUImage/RenderUIElement/RenderUIElement/ViewController.swift
//
//  ViewController.swift
//  RenderUIElement
//
//  Created by burt on 3/21/16.
//  Copyright © 2016 burt. All rights reserved.
//

import UIKit
import GPUImage

class ViewController: UIViewController {
    
    var filterView: GPUImageView!
    var videoCamera : GPUImageVideoCamera!
    var uiElement: GPUImageUIElement!
    var filter:GPUImageBrightnessFilter!
    var blendFilter: GPUImageAlphaBlendFilter!
    var uiElementInput: GPUImageUIElement!
    
    var beardImageView: UIImageView!
    var christmasHatImageView: UIImageView!
    var mustacheImageView: UIImageView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        filterView = GPUImageView(frame: self.view.bounds)
        self.view.addSubview(filterView)
        
        videoCamera = GPUImageVideoCamera(
            sessionPreset: AVCaptureSessionPreset640x480,
            cameraPosition: AVCaptureDevicePosition.Back)
        videoCamera.outputImageOrientation = .Portrait
        videoCamera.horizontallyMirrorFrontFacingCamera = false
        videoCamera.horizontallyMirrorRearFacingCamera = false
        
        filter = GPUImageBrightnessFilter()
        blendFilter = GPUImageAlphaBlendFilter()
        blendFilter.mix = 1.0
        
        videoCamera.addTarget(filter)
        
        // here I try to add a label as UIElement
        let contentView = UIView(frame: CGRect(x: 0, y: 0, width: 640, height: 480))
        self.beardImageView = UIImageView(image: UIImage(named: "beard")!)
        self.christmasHatImageView = UIImageView(image: UIImage(named: "christmas_hat"))
        self.mustacheImageView = UIImageView(image: UIImage(named: "mustache"))
        self.mustacheImageView.center = CGPointMake(100,100)
        
        contentView.addSubview(beardImageView)
        contentView.addSubview(christmasHatImageView)
        contentView.addSubview(mustacheImageView)
        
        uiElementInput = GPUImageUIElement(view: contentView)
        
        filter.addTarget(blendFilter)
        uiElementInput.addTarget(blendFilter)
        blendFilter.addTarget(filterView)
        
        filter.frameProcessingCompletionBlock = { filter, time in
            self.uiElementInput.update()
        }
        
        videoCamera.startCameraCapture()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}

